### ISSUE TITLE
USER Port access (especially for switch interface box and joystick Special Needs access)

### ISSUE THREAD

**Issue by oneswitch** [mattgodbolt/jsbeeb#160]
Adding USER PORT emulation would facilitate access to a lot of BBC Micro software designed for disabled people unable to use conventional controls. Typically this was via a switch interface box with four sockets, or a joystick. This link explains a little more from a Brilliant Computing catalogue.

[https://www.dropbox.com/s/idmzhdv84rmyr5v/09-10.jpg?dl=0](url)

I believe it's an 8-line interface, which writes the status to memory location ?&FE60 as follows....
​
?&FE60 = 255 11111111 = no switches pressed via the USER PORT
?&FE60 = 254 11111110 = Switch 1 / Right pressed via the USER PORT
?&FE60 = 253 11111101 = Switch 2 / Left pressed via the USER PORT 
?&FE60 = 251 11111011 = Switch 3 / Up pressed via the USER PORT
?&FE60 = 247 11110111 = Switch 4 / Down pressed via the USER PORT

.... and for more than one at a time, working as this example: 252 11111100 = Switch 1 + 2 together..... Quite likely that some activities may use a fifth switch for a fire button. 

Here's some software you can test this with....

Thurrock Care Switch programmes: If you select the talking word processor, you can test all 4 switches quite easily. https://bbc.godbolt.org/?autoboot&disc=https://dl.dropboxusercontent.com/s/8pralcrfztknabu/DISC000.ssd?dl=0

Joystick Games: Select USER PORT for connection. As you can see, these lock up because there's no way to reset the analogue or user ports. They both seem to have values jamming them on. https://bbc.godbolt.org/?autoboot&disc=https://dl.dropboxusercontent.com/s/zetb1jcf9julkmq/DISC013.ssd?dl=0

Spot on Games: on-line. Also locks up before you can start it... https://bbc.godbolt.org/?autoboot&disc=https://dl.dropboxusercontent.com/s/rz1s4i2j3vtnvhx/DISC017.ssd?dl=0

It may be necessary to set the user-port to 255 (no switches pressed) and also to "reset" the analogue port to get stuff all working, as some titles will lock up before starting if they think you are holding a control down.

Fingers crossed for one day.

**Comment by oneswitch** [comment]
Have a look at this for a bit more on the software this would open up again: [https://www.youtube.com/watch?v=PNNIG17tQVI&t=117s](url)

**Comment by oneswitch** [comment]
Would still absolutely love to see this working one day, e.g. for this: https://oneswitch.org.uk/jsbeeb/?autoboot&disc1=(Brilliant_Computing)_Joystick_Games_1_(1986).ssd

I may have a try myself.

**Comment by oneswitch** [comment]
Seems like free AI coding tools are well out of their depth, but they/we had a fair crack at it.

To my mind, this might work well on a lot of the old BBC Micro accessibility games and activities for accessibility switch users, perhaps as a hardware option to turn on Accessibility Switches.

**SWITCH 1** 
INPUT: F1/Mouse-left-click
OUTPUT: ADVAL=1 (analogue port left-switch) + ?&FE60=254 (user-port switch 1/joystick →) + INKEY=112 (F1)

**SWITCH 2**
INPUT: F2/Mouse-right-click
OUTPUT: ADVAL=2 (analogue port right-switch) + ?&FE60=253 (user-port switch 2/joystick ←) + INKEY=113 (F2)

**SWITCH 3**
INPUT F3
OUTPUT: ?&FE60=251 (user-port switch 3/joystick ↑)  + INKEY=114 (F3)

**SWITCH 4**
INPUT: F4 
OUTPUT: ?&FE60=247 (user-port switch 4/joystick ↓), INKEY=115 (F4)

ADVAL = 0 / ?&FE60 = 255 = nothing pressed. The input controls would need to act in a momentary fashion (e.g. active until input is released). (You can get other values from ?&FE60 for if user-port inputs are chorded - but most games only react to one-switch at a time).

This is very rough (at the time of writing) but there are some useful test titles there: [OneSwitch BBC Micro page](https://www.oneswitch.org.uk/art.php?id=386).

Thanks again to you and all those who helped to bring JSbeeb into being.
